// angular2/es6/dev/src/core/compiler/view.js
import { ListWrapper, MapWrapper, Map, StringMapWrapper } from 'angular2/src/facade/collection';
import { Locals } from 'angular2/change_detection';
import { ElementBinder } from './element_binder';
import { isPresent, isBlank, BaseException } from 'angular2/src/facade/lang';
import { ViewRef, ProtoViewRef, internalView } from './view_ref';
export class AppProtoViewMergeMapping {
    constructor(renderProtoViewMergeMapping) {
        this.renderProtoViewRef = renderProtoViewMergeMapping.mergedProtoViewRef;
        this.renderFragmentCount = renderProtoViewMergeMapping.fragmentCount;
        this.renderElementIndices = renderProtoViewMergeMapping.mappedElementIndices;
        this.renderInverseElementIndices =
            inverseIndexMapping(this.renderElementIndices, this.renderElementIndices.length);
        this.renderTextIndices = renderProtoViewMergeMapping.mappedTextIndices;
        this.hostElementIndicesByViewIndex = renderProtoViewMergeMapping.hostElementIndicesByViewIndex;
        this.nestedViewIndicesByElementIndex =
            inverseIndexMapping(this.hostElementIndicesByViewIndex, this.renderElementIndices.length);
        this.nestedViewCountByViewIndex = renderProtoViewMergeMapping.nestedViewCountByViewIndex;
    }
}
function inverseIndexMapping(input, resultLength) {
    var result = ListWrapper.createFixedSize(resultLength);
    for (var i = 0; i < input.length; i++) {
        var value = input[i];
        if (isPresent(value)) {
            result[input[i]] = i;
        }
    }
    return result;
}
export class AppViewContainer {
    constructor() {
        // The order in this list matches the DOM order.
        this.views = [];
    }
}
/**
 * Cost of making objects: http://jsperf.com/instantiate-size-of-object
 *
 */
export class AppView {
    constructor(renderer, proto, mainMergeMapping, viewOffset, elementOffset, textOffset, protoLocals, render, renderFragment) {
        this.renderer = renderer;
        this.proto = proto;
        this.mainMergeMapping = mainMergeMapping;
        this.viewOffset = viewOffset;
        this.elementOffset = elementOffset;
        this.textOffset = textOffset;
        this.render = render;
        this.renderFragment = renderFragment;
        // AppViews that have been merged in depth first order.
        // This list is shared between all merged views. Use this.elementOffset to get the local
        // entries.
        this.views = null;
        // ElementInjectors of all AppViews in views grouped by view.
        // This list is shared between all merged views. Use this.elementOffset to get the local
        // entries.
        this.elementInjectors = null;
        // ViewContainers of all AppViews in views grouped by view.
        // This list is shared between all merged views. Use this.elementOffset to get the local
        // entries.
        this.viewContainers = null;
        // PreBuiltObjects of all AppViews in views grouped by view.
        // This list is shared between all merged views. Use this.elementOffset to get the local
        // entries.
        this.preBuiltObjects = null;
        this.changeDetector = null;
        /**
         * The context against which data-binding expressions in this view are evaluated against.
         * This is always a component instance.
         */
        this.context = null;
        this.ref = new ViewRef(this);
        this.locals = new Locals(null, MapWrapper.clone(protoLocals)); // TODO optimize this
    }
    init(changeDetector, elementInjectors, rootElementInjectors, preBuiltObjects, views, elementRefs, viewContainers) {
        this.changeDetector = changeDetector;
        this.elementInjectors = elementInjectors;
        this.rootElementInjectors = rootElementInjectors;
        this.preBuiltObjects = preBuiltObjects;
        this.views = views;
        this.elementRefs = elementRefs;
        this.viewContainers = viewContainers;
    }
    setLocal(contextName, value) {
        if (!this.hydrated())
            throw new BaseException('Cannot set locals on dehydrated view.');
        if (!this.proto.variableBindings.has(contextName)) {
            return;
        }
        var templateName = this.proto.variableBindings.get(contextName);
        this.locals.set(templateName, value);
    }
    hydrated() { return isPresent(this.context); }
    /**
     * Triggers the event handlers for the element and the directives.
     *
     * This method is intended to be called from directive EventEmitters.
     *
     * @param {string} eventName
     * @param {*} eventObj
     * @param {int} boundElementIndex
     */
    triggerEventHandlers(eventName, eventObj, boundElementIndex) {
        var locals = new Map();
        locals.set('$event', eventObj);
        this.dispatchEvent(boundElementIndex, eventName, locals);
    }
    // dispatch to element injector or text nodes based on context
    notifyOnBinding(b, currentValue) {
        if (b.isTextNode()) {
            this.renderer.setText(this.render, this.mainMergeMapping.renderTextIndices[b.elementIndex + this.textOffset], currentValue);
        }
        else {
            var elementRef = this.elementRefs[this.elementOffset + b.elementIndex];
            if (b.isElementProperty()) {
                this.renderer.setElementProperty(elementRef, b.propertyName, currentValue);
            }
            else if (b.isElementAttribute()) {
                this.renderer.setElementAttribute(elementRef, b.propertyName, currentValue);
            }
            else if (b.isElementClass()) {
                this.renderer.setElementClass(elementRef, b.propertyName, currentValue);
            }
            else if (b.isElementStyle()) {
                var unit = isPresent(b.propertyUnit) ? b.propertyUnit : '';
                this.renderer.setElementStyle(elementRef, b.propertyName, `${currentValue}${unit}`);
            }
            else {
                throw new BaseException('Unsupported directive record');
            }
        }
    }
    notifyOnAllChangesDone() {
        var eiCount = this.proto.elementBinders.length;
        var ei = this.elementInjectors;
        for (var i = eiCount - 1; i >= 0; i--) {
            if (isPresent(ei[i + this.elementOffset]))
                ei[i + this.elementOffset].onAllChangesDone();
        }
    }
    getDirectiveFor(directive) {
        var elementInjector = this.elementInjectors[this.elementOffset + directive.elementIndex];
        return elementInjector.getDirectiveAtIndex(directive.directiveIndex);
    }
    getNestedView(boundElementIndex) {
        var viewIndex = this.mainMergeMapping.nestedViewIndicesByElementIndex[boundElementIndex];
        return isPresent(viewIndex) ? this.views[viewIndex] : null;
    }
    getDetectorFor(directive) {
        var childView = this.getNestedView(this.elementOffset + directive.elementIndex);
        return isPresent(childView) ? childView.changeDetector : null;
    }
    invokeElementMethod(elementIndex, methodName, args) {
        this.renderer.invokeElementMethod(this.elementRefs[elementIndex], methodName, args);
    }
    // implementation of RenderEventDispatcher#dispatchRenderEvent
    dispatchRenderEvent(renderElementIndex, eventName, locals) {
        var elementRef = this.elementRefs[this.mainMergeMapping.renderInverseElementIndices[renderElementIndex]];
        var view = internalView(elementRef.parentView);
        return view.dispatchEvent(elementRef.boundElementIndex, eventName, locals);
    }
    // returns false if preventDefault must be applied to the DOM event
    dispatchEvent(boundElementIndex, eventName, locals) {
        // Most of the time the event will be fired only when the view is in the live document.
        // However, in a rare circumstance the view might get dehydrated, in between the event
        // queuing up and firing.
        var allowDefaultBehavior = true;
        if (this.hydrated()) {
            var elBinder = this.proto.elementBinders[boundElementIndex - this.elementOffset];
            if (isBlank(elBinder.hostListeners))
                return allowDefaultBehavior;
            var eventMap = elBinder.hostListeners[eventName];
            if (isBlank(eventMap))
                return allowDefaultBehavior;
            MapWrapper.forEach(eventMap, (expr, directiveIndex) => {
                var context;
                if (directiveIndex === -1) {
                    context = this.context;
                }
                else {
                    context = this.elementInjectors[boundElementIndex].getDirectiveAtIndex(directiveIndex);
                }
                var result = expr.eval(context, new Locals(this.locals, locals));
                if (isPresent(result)) {
                    allowDefaultBehavior = allowDefaultBehavior && result == true;
                }
            });
        }
        return allowDefaultBehavior;
    }
}
/**
 *
 */
export class AppProtoView {
    constructor(type, isEmbeddedFragment, render, protoChangeDetector, variableBindings, variableLocations, textBindingCount) {
        this.type = type;
        this.isEmbeddedFragment = isEmbeddedFragment;
        this.render = render;
        this.protoChangeDetector = protoChangeDetector;
        this.variableBindings = variableBindings;
        this.variableLocations = variableLocations;
        this.textBindingCount = textBindingCount;
        this.elementBinders = [];
        this.protoLocals = new Map();
        this.isRecursive = null;
        this.ref = new ProtoViewRef(this);
        if (isPresent(variableBindings)) {
            MapWrapper.forEach(variableBindings, (templateName, _) => { this.protoLocals.set(templateName, null); });
        }
    }
    bindElement(parent, distanceToParent, protoElementInjector, componentDirective = null) {
        var elBinder = new ElementBinder(this.elementBinders.length, parent, distanceToParent, protoElementInjector, componentDirective);
        this.elementBinders.push(elBinder);
        return elBinder;
    }
    /**
     * Adds an event binding for the last created ElementBinder via bindElement.
     *
     * If the directive index is a positive integer, the event is evaluated in the context of
     * the given directive.
     *
     * If the directive index is -1, the event is evaluated in the context of the enclosing view.
     *
     * @param {string} eventName
     * @param {AST} expression
     * @param {int} directiveIndex The directive index in the binder or -1 when the event is not bound
     *                             to a directive
     */
    bindEvent(eventBindings, boundElementIndex, directiveIndex = -1) {
        var elBinder = this.elementBinders[boundElementIndex];
        var events = elBinder.hostListeners;
        if (isBlank(events)) {
            events = StringMapWrapper.create();
            elBinder.hostListeners = events;
        }
        for (var i = 0; i < eventBindings.length; i++) {
            var eventBinding = eventBindings[i];
            var eventName = eventBinding.fullName;
            var event = StringMapWrapper.get(events, eventName);
            if (isBlank(event)) {
                event = new Map();
                StringMapWrapper.set(events, eventName, event);
            }
            event.set(directiveIndex, eventBinding.source);
        }
    }
}
