// angular2/es6/dev/src/directives/ng_style.js
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") return Reflect.decorate(decorators, target, key, desc);
    switch (arguments.length) {
        case 2: return decorators.reduceRight(function(o, d) { return (d && d(o)) || o; }, target);
        case 3: return decorators.reduceRight(function(o, d) { return (d && d(target, key)), void 0; }, void 0);
        case 4: return decorators.reduceRight(function(o, d) { return (d && d(target, key, o)) || o; }, desc);
    }
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
import { Directive, LifecycleEvent } from 'angular2/annotations';
import { ElementRef } from 'angular2/core';
import { Pipes } from 'angular2/src/change_detection/pipes/pipes';
import { isPresent } from 'angular2/src/facade/lang';
import { Renderer } from 'angular2/src/render/api';
/**
 * Adds or removes styles based on an {expression}.
 *
 * When the expression assigned to `ng-style` evaluates to an object, the corresponding element
 * styles are updated. Style names to update are taken from the object keys and values - from the
 * corresponding object values.
 *
 * # Example:
 *
 * ```
 * <div ng-style="{'text-align': alignEpr}"></div>
 * ```
 *
 * In the above example the `text-align` style will be updated based on the `alignEpr` value
 * changes.
 *
 * # Syntax
 *
 * - `<div ng-style="{'text-align': alignEpr}"></div>`
 * - `<div ng-style="styleExp"></div>`
 */
export let NgStyle = class {
    constructor(_pipes, _ngEl, _renderer) {
        this._pipes = _pipes;
        this._ngEl = _ngEl;
        this._renderer = _renderer;
    }
    set rawStyle(v) {
        this._rawStyle = v;
        this._pipe = this._pipes.get('keyValDiff', this._rawStyle);
    }
    onCheck() {
        var diff = this._pipe.transform(this._rawStyle, null);
        if (isPresent(diff) && isPresent(diff.wrapped)) {
            this._applyChanges(diff.wrapped);
        }
    }
    _applyChanges(diff) {
        diff.forEachAddedItem((record) => { this._setStyle(record.key, record.currentValue); });
        diff.forEachChangedItem((record) => { this._setStyle(record.key, record.currentValue); });
        diff.forEachRemovedItem((record) => { this._setStyle(record.key, null); });
    }
    _setStyle(name, val) {
        this._renderer.setElementStyle(this._ngEl, name, val);
    }
};
NgStyle = __decorate([
    Directive({
        selector: '[ng-style]',
        lifecycle: [LifecycleEvent.onCheck],
        properties: ['rawStyle: ng-style']
    }), 
    __metadata('design:paramtypes', [Pipes, ElementRef, Renderer])
], NgStyle);
