// angular2/src/web-workers/worker/application.js
'use strict';var lang_1 = require("angular2/src/facade/lang");
/**
 * Bootstrapping a Webworker Application
 *
 * You instantiate the application side by calling bootstrapWebworker from your webworker index
 * script.
 * You can call bootstrapWebworker() exactly as you would call bootstrap() in a regular Angular
 * application
 * See the bootstrap() docs for more details.
 */
function bootstrapWebworker(appComponentType, componentInjectableBindings, errorReporter) {
    if (componentInjectableBindings === void 0) { componentInjectableBindings = null; }
    if (errorReporter === void 0) { errorReporter = null; }
    throw new lang_1.BaseException("Not Implemented");
}
exports.bootstrapWebworker = bootstrapWebworker;
var WorkerMessageBus = (function () {
    function WorkerMessageBus(sink, source) {
        this.sink = sink;
        this.source = source;
    }
    return WorkerMessageBus;
})();
exports.WorkerMessageBus = WorkerMessageBus;
var WorkerMessageBusSink = (function () {
    function WorkerMessageBusSink() {
    }
    WorkerMessageBusSink.prototype.send = function (message) { postMessage(message, null); };
    return WorkerMessageBusSink;
})();
exports.WorkerMessageBusSink = WorkerMessageBusSink;
var WorkerMessageBusSource = (function () {
    function WorkerMessageBusSource() {
    }
    WorkerMessageBusSource.prototype.listen = function (fn) { addEventListener("message", fn); };
    return WorkerMessageBusSource;
})();
exports.WorkerMessageBusSource = WorkerMessageBusSource;
