// angular2/src/core/compiler/compiler.js
'use strict';var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") return Reflect.decorate(decorators, target, key, desc);
    switch (arguments.length) {
        case 2: return decorators.reduceRight(function(o, d) { return (d && d(o)) || o; }, target);
        case 3: return decorators.reduceRight(function(o, d) { return (d && d(target, key)), void 0; }, void 0);
        case 4: return decorators.reduceRight(function(o, d) { return (d && d(target, key, o)) || o; }, desc);
    }
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
var di_1 = require('angular2/di');
var lang_1 = require('angular2/src/facade/lang');
var async_1 = require('angular2/src/facade/async');
var collection_1 = require('angular2/src/facade/collection');
var directive_resolver_1 = require('./directive_resolver');
var view_1 = require('./view');
var element_injector_1 = require('./element_injector');
var view_resolver_1 = require('./view_resolver');
var component_url_mapper_1 = require('./component_url_mapper');
var proto_view_factory_1 = require('./proto_view_factory');
var url_resolver_1 = require('angular2/src/services/url_resolver');
var app_root_url_1 = require('angular2/src/services/app_root_url');
var renderApi = require('angular2/src/render/api');
/**
 * Cache that stores the AppProtoView of the template of a component.
 * Used to prevent duplicate work and resolve cyclic dependencies.
 */
var CompilerCache = (function () {
    function CompilerCache() {
        this._cache = new collection_1.Map();
        this._hostCache = new collection_1.Map();
    }
    CompilerCache.prototype.set = function (component, protoView) { this._cache.set(component, protoView); };
    CompilerCache.prototype.get = function (component) {
        var result = this._cache.get(component);
        return lang_1.normalizeBlank(result);
    };
    CompilerCache.prototype.setHost = function (component, protoView) {
        this._hostCache.set(component, protoView);
    };
    CompilerCache.prototype.getHost = function (component) {
        var result = this._hostCache.get(component);
        return lang_1.normalizeBlank(result);
    };
    CompilerCache.prototype.clear = function () {
        this._cache.clear();
        this._hostCache.clear();
    };
    CompilerCache = __decorate([
        di_1.Injectable(), 
        __metadata('design:paramtypes', [])
    ], CompilerCache);
    return CompilerCache;
})();
exports.CompilerCache = CompilerCache;
/**
 *
 * ## URL Resolution
 *
 * ```
 * var appRootUrl: AppRootUrl = ...;
 * var componentUrlMapper: ComponentUrlMapper = ...;
 * var urlResolver: UrlResolver = ...;
 *
 * var componentType: Type = ...;
 * var componentAnnotation: ComponentAnnotation = ...;
 * var viewAnnotation: ViewAnnotation = ...;
 *
 * // Resolving a URL
 *
 * var url = viewAnnotation.templateUrl;
 * var componentUrl = componentUrlMapper.getUrl(componentType);
 * var componentResolvedUrl = urlResolver.resolve(appRootUrl.value, componentUrl);
 * var templateResolvedUrl = urlResolver.resolve(componetResolvedUrl, url);
 * ```
 */
var Compiler = (function () {
    /**
     * @private
     */
    function Compiler(reader, cache, viewResolver, componentUrlMapper, urlResolver, render, protoViewFactory, appUrl) {
        this._protoViewsToBeMerged = [];
        this._reader = reader;
        this._compilerCache = cache;
        this._compiling = new collection_1.Map();
        this._viewResolver = viewResolver;
        this._componentUrlMapper = componentUrlMapper;
        this._urlResolver = urlResolver;
        this._appUrl = appUrl.value;
        this._render = render;
        this._protoViewFactory = protoViewFactory;
    }
    Compiler.prototype._bindDirective = function (directiveTypeOrBinding) {
        if (directiveTypeOrBinding instanceof element_injector_1.DirectiveBinding) {
            return directiveTypeOrBinding;
        }
        else if (directiveTypeOrBinding instanceof di_1.Binding) {
            var annotation = this._reader.resolve(directiveTypeOrBinding.token);
            return element_injector_1.DirectiveBinding.createFromBinding(directiveTypeOrBinding, annotation);
        }
        else {
            var annotation = this._reader.resolve(directiveTypeOrBinding);
            return element_injector_1.DirectiveBinding.createFromType(directiveTypeOrBinding, annotation);
        }
    };
    // Create a hostView as if the compiler encountered <hostcmp></hostcmp>.
    // Used for bootstrapping.
    Compiler.prototype.compileInHost = function (componentTypeOrBinding) {
        var _this = this;
        var componentType = lang_1.isType(componentTypeOrBinding) ? componentTypeOrBinding :
            componentTypeOrBinding.token;
        var hostAppProtoView = this._compilerCache.getHost(componentType);
        var hostPvPromise;
        if (lang_1.isPresent(hostAppProtoView)) {
            hostPvPromise = async_1.PromiseWrapper.resolve(hostAppProtoView);
        }
        else {
            var componentBinding = this._bindDirective(componentTypeOrBinding);
            Compiler._assertTypeIsComponent(componentBinding);
            var directiveMetadata = componentBinding.metadata;
            hostPvPromise =
                this._render.compileHost(directiveMetadata)
                    .then(function (hostRenderPv) {
                    var protoView = _this._protoViewFactory.createAppProtoViews(componentBinding, hostRenderPv, [componentBinding]);
                    _this._compilerCache.setHost(componentType, protoView);
                    return _this._compileNestedProtoViews(hostRenderPv, protoView, componentType);
                });
        }
        return hostPvPromise.then(function (hostAppProtoView) {
            return _this._mergeUnmergedProtoViews().then(function (_) { return hostAppProtoView.ref; });
        });
    };
    Compiler.prototype._mergeUnmergedProtoViews = function () {
        var _this = this;
        var protoViewsToBeMerged = this._protoViewsToBeMerged;
        this._protoViewsToBeMerged = [];
        return async_1.PromiseWrapper.all(protoViewsToBeMerged.map(function (appProtoView) {
            return _this._render.mergeProtoViewsRecursively(_this._collectMergeRenderProtoViews(appProtoView))
                .then(function (mergeResult) {
                appProtoView.mergeMapping = new view_1.AppProtoViewMergeMapping(mergeResult);
            });
        }));
    };
    Compiler.prototype._collectMergeRenderProtoViews = function (appProtoView) {
        var result = [appProtoView.render];
        for (var i = 0; i < appProtoView.elementBinders.length; i++) {
            var binder = appProtoView.elementBinders[i];
            if (lang_1.isPresent(binder.nestedProtoView)) {
                if (binder.hasStaticComponent() ||
                    (binder.hasEmbeddedProtoView() && binder.nestedProtoView.isEmbeddedFragment)) {
                    result.push(this._collectMergeRenderProtoViews(binder.nestedProtoView));
                }
                else {
                    result.push(null);
                }
            }
        }
        return result;
    };
    Compiler.prototype._compile = function (componentBinding) {
        var _this = this;
        var component = componentBinding.key.token;
        var protoView = this._compilerCache.get(component);
        if (lang_1.isPresent(protoView)) {
            // The component has already been compiled into an AppProtoView,
            // returns a plain AppProtoView, not wrapped inside of a Promise.
            // Needed for recursive components.
            return protoView;
        }
        var resultPromise = this._compiling.get(component);
        if (lang_1.isPresent(resultPromise)) {
            // The component is already being compiled, attach to the existing Promise
            // instead of re-compiling the component.
            // It happens when a template references a component multiple times.
            return resultPromise;
        }
        var view = this._viewResolver.resolve(component);
        var directives = this._flattenDirectives(view);
        for (var i = 0; i < directives.length; i++) {
            if (!Compiler._isValidDirective(directives[i])) {
                throw new lang_1.BaseException("Unexpected directive value '" + lang_1.stringify(directives[i]) + "' on the View of component '" + lang_1.stringify(component) + "'");
            }
        }
        var boundDirectives = this._removeDuplicatedDirectives(collection_1.ListWrapper.map(directives, function (directive) { return _this._bindDirective(directive); }));
        var renderTemplate = this._buildRenderTemplate(component, view, boundDirectives);
        resultPromise = this._render.compile(renderTemplate)
            .then(function (renderPv) {
            var protoView = _this._protoViewFactory.createAppProtoViews(componentBinding, renderPv, boundDirectives);
            // Populate the cache before compiling the nested components,
            // so that components can reference themselves in their template.
            _this._compilerCache.set(component, protoView);
            collection_1.MapWrapper.delete(_this._compiling, component);
            return _this._compileNestedProtoViews(renderPv, protoView, component);
        });
        this._compiling.set(component, resultPromise);
        return resultPromise;
    };
    Compiler.prototype._removeDuplicatedDirectives = function (directives) {
        var directivesMap = new collection_1.Map();
        directives.forEach(function (dirBinding) { directivesMap.set(dirBinding.key.id, dirBinding); });
        return collection_1.MapWrapper.values(directivesMap);
    };
    Compiler.prototype._compileNestedProtoViews = function (renderProtoView, appProtoView, componentType) {
        var _this = this;
        var nestedPVPromises = [];
        this._loopComponentElementBinders(appProtoView, function (parentPv, elementBinder) {
            var nestedComponent = elementBinder.componentDirective;
            var elementBinderDone = function (nestedPv) { elementBinder.nestedProtoView = nestedPv; };
            var nestedCall = _this._compile(nestedComponent);
            if (lang_1.isPromise(nestedCall)) {
                nestedPVPromises.push(nestedCall.then(elementBinderDone));
            }
            else {
                elementBinderDone(nestedCall);
            }
        });
        return async_1.PromiseWrapper.all(nestedPVPromises)
            .then(function (_) {
            _this._collectMergableProtoViews(appProtoView, componentType);
            return appProtoView;
        });
    };
    Compiler.prototype._collectMergableProtoViews = function (appProtoView, componentType) {
        var isRecursive = false;
        for (var i = 0; i < appProtoView.elementBinders.length; i++) {
            var binder = appProtoView.elementBinders[i];
            if (binder.hasStaticComponent()) {
                if (lang_1.isBlank(binder.nestedProtoView.isRecursive)) {
                    // cycle via a component. We are in the tail recursion,
                    // so all components should have their isRecursive flag set already.
                    isRecursive = true;
                    break;
                }
            }
            else if (binder.hasEmbeddedProtoView()) {
                this._collectMergableProtoViews(binder.nestedProtoView, componentType);
            }
        }
        if (isRecursive) {
            if (appProtoView.isEmbeddedFragment) {
                throw new lang_1.BaseException("<ng-content> is used within the recursive path of " + lang_1.stringify(componentType));
            }
            if (appProtoView.type === renderApi.ViewType.COMPONENT) {
                throw new lang_1.BaseException("Unconditional component cycle in " + lang_1.stringify(componentType));
            }
        }
        if (appProtoView.type === renderApi.ViewType.EMBEDDED ||
            appProtoView.type === renderApi.ViewType.HOST) {
            this._protoViewsToBeMerged.push(appProtoView);
        }
        appProtoView.isRecursive = isRecursive;
    };
    Compiler.prototype._loopComponentElementBinders = function (appProtoView, callback) {
        var _this = this;
        appProtoView.elementBinders.forEach(function (elementBinder) {
            if (lang_1.isPresent(elementBinder.componentDirective)) {
                callback(appProtoView, elementBinder);
            }
            else if (lang_1.isPresent(elementBinder.nestedProtoView)) {
                _this._loopComponentElementBinders(elementBinder.nestedProtoView, callback);
            }
        });
    };
    Compiler.prototype._buildRenderTemplate = function (component, view, directives) {
        var _this = this;
        var componentUrl = this._urlResolver.resolve(this._appUrl, this._componentUrlMapper.getUrl(component));
        var templateAbsUrl = null;
        var styleAbsUrls = null;
        if (lang_1.isPresent(view.templateUrl)) {
            templateAbsUrl = this._urlResolver.resolve(componentUrl, view.templateUrl);
        }
        else if (lang_1.isPresent(view.template)) {
            // Note: If we have an inline template, we also need to send
            // the url for the component to the render so that it
            // is able to resolve urls in stylesheets.
            templateAbsUrl = componentUrl;
        }
        if (lang_1.isPresent(view.styleUrls)) {
            styleAbsUrls =
                collection_1.ListWrapper.map(view.styleUrls, function (url) { return _this._urlResolver.resolve(componentUrl, url); });
        }
        return new renderApi.ViewDefinition({
            componentId: lang_1.stringify(component),
            templateAbsUrl: templateAbsUrl, template: view.template,
            styleAbsUrls: styleAbsUrls,
            styles: view.styles,
            directives: collection_1.ListWrapper.map(directives, function (directiveBinding) { return directiveBinding.metadata; })
        });
    };
    Compiler.prototype._flattenDirectives = function (template) {
        if (lang_1.isBlank(template.directives))
            return [];
        var directives = [];
        this._flattenList(template.directives, directives);
        return directives;
    };
    Compiler.prototype._flattenList = function (tree, out) {
        for (var i = 0; i < tree.length; i++) {
            var item = di_1.resolveForwardRef(tree[i]);
            if (lang_1.isArray(item)) {
                this._flattenList(item, out);
            }
            else {
                out.push(item);
            }
        }
    };
    Compiler._isValidDirective = function (value) {
        return lang_1.isPresent(value) && (value instanceof lang_1.Type || value instanceof di_1.Binding);
    };
    Compiler._assertTypeIsComponent = function (directiveBinding) {
        if (directiveBinding.metadata.type !== renderApi.DirectiveMetadata.COMPONENT_TYPE) {
            throw new lang_1.BaseException("Could not load '" + lang_1.stringify(directiveBinding.key.token) + "' because it is not a component.");
        }
    };
    Compiler = __decorate([
        di_1.Injectable(), 
        __metadata('design:paramtypes', [directive_resolver_1.DirectiveResolver, CompilerCache, view_resolver_1.ViewResolver, component_url_mapper_1.ComponentUrlMapper, url_resolver_1.UrlResolver, renderApi.RenderCompiler, proto_view_factory_1.ProtoViewFactory, app_root_url_1.AppRootUrl])
    ], Compiler);
    return Compiler;
})();
exports.Compiler = Compiler;
