// angular2/es6/prod/src/web-workers/shared/render_view_with_fragments_store.js
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") return Reflect.decorate(decorators, target, key, desc);
    switch (arguments.length) {
        case 2: return decorators.reduceRight(function(o, d) { return (d && d(o)) || o; }, target);
        case 3: return decorators.reduceRight(function(o, d) { return (d && d(target, key)), void 0; }, void 0);
        case 4: return decorators.reduceRight(function(o, d) { return (d && d(target, key, o)) || o; }, desc);
    }
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
var __param = (this && this.__param) || function (paramIndex, decorator) {
    return function (target, key) { decorator(target, key, paramIndex); }
};
import { Injectable, Inject } from "angular2/di";
import { RenderViewRef, RenderFragmentRef, RenderViewWithFragments } from "angular2/src/render/api";
import { ON_WEBWORKER } from "angular2/src/web-workers/shared/api";
import { ListWrapper } from "angular2/src/facade/collection";
export let RenderViewWithFragmentsStore = class {
    constructor(onWebWorker) {
        this._nextIndex = 0;
        this._onWebWorker = onWebWorker;
        if (!onWebWorker) {
            this._lookupByIndex = new Map();
            this._lookupByView = new Map();
        }
    }
    allocate(fragmentCount) {
        var viewRef = new WorkerRenderViewRef(this._nextIndex++);
        var fragmentRefs = ListWrapper.createGrowableSize(fragmentCount);
        for (var i = 0; i < fragmentCount; i++) {
            fragmentRefs[i] = new WorkerRenderFragmentRef(this._nextIndex++);
        }
        return new RenderViewWithFragments(viewRef, fragmentRefs);
    }
    store(view, startIndex) {
        this._lookupByIndex.set(startIndex, view.viewRef);
        this._lookupByView.set(view.viewRef, startIndex);
        startIndex++;
        ListWrapper.forEach(view.fragmentRefs, (ref) => {
            this._lookupByIndex.set(startIndex, ref);
            this._lookupByView.set(ref, startIndex);
            startIndex++;
        });
    }
    retreive(ref) {
        if (ref == null) {
            return null;
        }
        return this._lookupByIndex.get(ref);
    }
    serializeRenderViewRef(viewRef) {
        return this._serializeRenderFragmentOrViewRef(viewRef);
    }
    serializeRenderFragmentRef(fragmentRef) {
        return this._serializeRenderFragmentOrViewRef(fragmentRef);
    }
    deserializeRenderViewRef(ref) {
        if (ref == null) {
            return null;
        }
        if (this._onWebWorker) {
            return WorkerRenderViewRef.deserialize(ref);
        }
        else {
            return this.retreive(ref);
        }
    }
    deserializeRenderFragmentRef(ref) {
        if (ref == null) {
            return null;
        }
        if (this._onWebWorker) {
            return WorkerRenderFragmentRef.deserialize(ref);
        }
        else {
            return this.retreive(ref);
        }
    }
    _serializeRenderFragmentOrViewRef(ref) {
        if (ref == null) {
            return null;
        }
        if (this._onWebWorker) {
            return ref.serialize();
        }
        else {
            return this._lookupByView.get(ref);
        }
    }
    serializeViewWithFragments(view) {
        if (view == null) {
            return null;
        }
        if (this._onWebWorker) {
            return {
                'viewRef': view.viewRef.serialize(),
                'fragmentRefs': ListWrapper.map(view.fragmentRefs, (val) => val.serialize())
            };
        }
        else {
            return {
                'viewRef': this._lookupByView.get(view.viewRef),
                'fragmentRefs': ListWrapper.map(view.fragmentRefs, (val) => this._lookupByView.get(val))
            };
        }
    }
    deserializeViewWithFragments(obj) {
        if (obj == null) {
            return null;
        }
        var viewRef;
        var fragments;
        if (this._onWebWorker) {
            viewRef = WorkerRenderViewRef.deserialize(obj['viewRef']);
            fragments =
                ListWrapper.map(obj['fragmentRefs'], (val) => WorkerRenderFragmentRef.deserialize(val));
            return new RenderViewWithFragments(viewRef, fragments);
        }
        else {
            viewRef = this.retreive(obj['viewRef']);
            fragments = ListWrapper.map(obj['fragmentRefs'], (val) => this.retreive(val));
            return new RenderViewWithFragments(viewRef, fragments);
        }
    }
};
RenderViewWithFragmentsStore = __decorate([
    Injectable(),
    __param(0, Inject(ON_WEBWORKER)), 
    __metadata('design:paramtypes', [Object])
], RenderViewWithFragmentsStore);
export class WorkerRenderViewRef extends RenderViewRef {
    constructor(refNumber) {
        super();
        this.refNumber = refNumber;
    }
    serialize() { return this.refNumber; }
    static deserialize(ref) { return new WorkerRenderViewRef(ref); }
}
export class WorkerRenderFragmentRef extends RenderFragmentRef {
    constructor(refNumber) {
        super();
        this.refNumber = refNumber;
    }
    serialize() { return this.refNumber; }
    static deserialize(ref) {
        return new WorkerRenderFragmentRef(ref);
    }
}
