// angular2/src/web-workers/ui/application.js
'use strict';var lang_1 = require("angular2/src/facade/lang");
/**
 * Bootstrapping a WebWorker
 *
 * You instantiate a WebWorker application by calling bootstrap with the URI of your worker's index
 * script
 * Note: The WebWorker script must call bootstrapWebworker once it is set up to complete the
 * bootstrapping process
 */
function bootstrap(uri) {
    throw new lang_1.BaseException("Not Implemented");
}
exports.bootstrap = bootstrap;
function spawnWorker(uri) {
    var worker = new Worker(uri);
    return new UIMessageBus(new UIMessageBusSink(worker), new UIMessageBusSource(worker));
}
exports.spawnWorker = spawnWorker;
var UIMessageBus = (function () {
    function UIMessageBus(sink, source) {
        this.sink = sink;
        this.source = source;
    }
    return UIMessageBus;
})();
exports.UIMessageBus = UIMessageBus;
var UIMessageBusSink = (function () {
    function UIMessageBusSink(_worker) {
        this._worker = _worker;
    }
    UIMessageBusSink.prototype.send = function (message) { this._worker.postMessage(message); };
    return UIMessageBusSink;
})();
exports.UIMessageBusSink = UIMessageBusSink;
var UIMessageBusSource = (function () {
    function UIMessageBusSource(_worker) {
        this._worker = _worker;
    }
    UIMessageBusSource.prototype.listen = function (fn) { this._worker.addEventListener("message", fn); };
    return UIMessageBusSource;
})();
exports.UIMessageBusSource = UIMessageBusSource;
