// angular2/es6/prod/src/render/dom/compiler/compiler.js
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") return Reflect.decorate(decorators, target, key, desc);
    switch (arguments.length) {
        case 2: return decorators.reduceRight(function(o, d) { return (d && d(o)) || o; }, target);
        case 3: return decorators.reduceRight(function(o, d) { return (d && d(target, key)), void 0; }, void 0);
        case 4: return decorators.reduceRight(function(o, d) { return (d && d(target, key, o)) || o; }, desc);
    }
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
var __param = (this && this.__param) || function (paramIndex, decorator) {
    return function (target, key) { decorator(target, key, paramIndex); }
};
import { Injectable } from 'angular2/di';
import { PromiseWrapper } from 'angular2/src/facade/async';
import { BaseException } from 'angular2/src/facade/lang';
import { DOM } from 'angular2/src/dom/dom_adapter';
import { ViewDefinition, ViewType, RenderCompiler, ViewEncapsulation } from '../../api';
import { CompilePipeline } from './compile_pipeline';
import { ViewLoader, TemplateAndStyles } from 'angular2/src/render/dom/compiler/view_loader';
import { DefaultStepFactory } from './compile_step_factory';
import { Parser } from 'angular2/src/change_detection/change_detection';
import * as pvm from '../view/proto_view_merger';
import { APP_ID_TOKEN } from '../dom_tokens';
import { Inject } from 'angular2/di';
import { SharedStylesHost } from '../view/shared_styles_host';
import { prependAll } from '../util';
/**
 * The compiler loads and translates the html templates of components into
 * nested ProtoViews. To decompose its functionality it uses
 * the CompilePipeline and the CompileSteps.
 */
export class DomCompiler extends RenderCompiler {
    constructor(_stepFactory, _viewLoader, _sharedStylesHost) {
        super();
        this._stepFactory = _stepFactory;
        this._viewLoader = _viewLoader;
        this._sharedStylesHost = _sharedStylesHost;
    }
    compile(view) {
        var tplPromise = this._viewLoader.load(view);
        return PromiseWrapper.then(tplPromise, (tplAndStyles) => this._compileView(view, tplAndStyles, ViewType.COMPONENT), (e) => {
            throw new BaseException(`Failed to load the template for "${view.componentId}" : ${e}`);
            return null;
        });
    }
    compileHost(directiveMetadata) {
        var hostViewDef = new ViewDefinition({
            componentId: directiveMetadata.id,
            templateAbsUrl: null, template: null,
            styles: null,
            styleAbsUrls: null,
            directives: [directiveMetadata],
            encapsulation: ViewEncapsulation.NONE
        });
        return this._compileView(hostViewDef, new TemplateAndStyles(`<${directiveMetadata.selector}></${directiveMetadata.selector}>`, []), ViewType.HOST);
    }
    mergeProtoViewsRecursively(protoViewRefs) {
        return PromiseWrapper.resolve(pvm.mergeProtoViewsRecursively(protoViewRefs));
    }
    _compileView(viewDef, templateAndStyles, protoViewType) {
        if (viewDef.encapsulation === ViewEncapsulation.EMULATED &&
            templateAndStyles.styles.length === 0) {
            viewDef = this._normalizeViewEncapsulationIfThereAreNoStyles(viewDef);
        }
        var pipeline = new CompilePipeline(this._stepFactory.createSteps(viewDef));
        var compiledStyles = pipeline.processStyles(templateAndStyles.styles);
        var compileElements = pipeline.processElements(DOM.createTemplate(templateAndStyles.template), protoViewType, viewDef);
        if (viewDef.encapsulation === ViewEncapsulation.NATIVE) {
            prependAll(DOM.content(compileElements[0].element), compiledStyles.map(style => DOM.createStyleElement(style)));
        }
        else {
            this._sharedStylesHost.addStyles(compiledStyles);
        }
        return PromiseWrapper.resolve(compileElements[0].inheritedProtoView.build());
    }
    _normalizeViewEncapsulationIfThereAreNoStyles(viewDef) {
        if (viewDef.encapsulation === ViewEncapsulation.EMULATED) {
            return new ViewDefinition({
                componentId: viewDef.componentId,
                templateAbsUrl: viewDef.templateAbsUrl, template: viewDef.template,
                styleAbsUrls: viewDef.styleAbsUrls,
                styles: viewDef.styles,
                directives: viewDef.directives,
                encapsulation: ViewEncapsulation.NONE
            });
        }
        else {
            return viewDef;
        }
    }
}
export let DefaultDomCompiler = class extends DomCompiler {
    constructor(parser, viewLoader, sharedStylesHost, appId) {
        super(new DefaultStepFactory(parser, appId), viewLoader, sharedStylesHost);
    }
};
DefaultDomCompiler = __decorate([
    Injectable(),
    __param(3, Inject(APP_ID_TOKEN)), 
    __metadata('design:paramtypes', [Parser, ViewLoader, SharedStylesHost, Object])
], DefaultDomCompiler);
