// angular2/es6/dev/src/web-workers/ui/application.js
import { BaseException } from "angular2/src/facade/lang";
/**
 * Bootstrapping a WebWorker
 *
 * You instantiate a WebWorker application by calling bootstrap with the URI of your worker's index
 * script
 * Note: The WebWorker script must call bootstrapWebworker once it is set up to complete the
 * bootstrapping process
 */
export function bootstrap(uri) {
    throw new BaseException("Not Implemented");
}
export function spawnWorker(uri) {
    var worker = new Worker(uri);
    return new UIMessageBus(new UIMessageBusSink(worker), new UIMessageBusSource(worker));
}
export class UIMessageBus {
    constructor(sink, source) {
        this.sink = sink;
        this.source = source;
    }
}
export class UIMessageBusSink {
    constructor(_worker) {
        this._worker = _worker;
    }
    send(message) { this._worker.postMessage(message); }
}
export class UIMessageBusSource {
    constructor(_worker) {
        this._worker = _worker;
    }
    listen(fn) { this._worker.addEventListener("message", fn); }
}
