// angular2/pipes.js
'use strict';/**
 * @module
 * @description
 * This module provides advanced support for extending change detection.
 */
var promise_pipe_1 = require('./src/change_detection/pipes/promise_pipe');
exports.PromisePipe = promise_pipe_1.PromisePipe;
var uppercase_pipe_1 = require('./src/change_detection/pipes/uppercase_pipe');
exports.UpperCasePipe = uppercase_pipe_1.UpperCasePipe;
var lowercase_pipe_1 = require('./src/change_detection/pipes/lowercase_pipe');
exports.LowerCasePipe = lowercase_pipe_1.LowerCasePipe;
var observable_pipe_1 = require('./src/change_detection/pipes/observable_pipe');
exports.ObservablePipe = observable_pipe_1.ObservablePipe;
var json_pipe_1 = require('./src/change_detection/pipes/json_pipe');
exports.JsonPipe = json_pipe_1.JsonPipe;
var iterable_changes_1 = require('./src/change_detection/pipes/iterable_changes');
exports.IterableChanges = iterable_changes_1.IterableChanges;
var keyvalue_changes_1 = require('./src/change_detection/pipes/keyvalue_changes');
exports.KeyValueChanges = keyvalue_changes_1.KeyValueChanges;
var date_pipe_1 = require('./src/change_detection/pipes/date_pipe');
exports.DatePipe = date_pipe_1.DatePipe;
var number_pipe_1 = require('./src/change_detection/pipes/number_pipe');
exports.DecimalPipe = number_pipe_1.DecimalPipe;
exports.PercentPipe = number_pipe_1.PercentPipe;
exports.CurrencyPipe = number_pipe_1.CurrencyPipe;
var limit_to_pipe_1 = require('./src/change_detection/pipes/limit_to_pipe');
exports.LimitToPipe = limit_to_pipe_1.LimitToPipe;
