// angular2/src/render/dom/compiler/compiler.js
'use strict';var __extends = (this && this.__extends) || function (d, b) {
    for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p];
    function __() { this.constructor = d; }
    __.prototype = b.prototype;
    d.prototype = new __();
};
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") return Reflect.decorate(decorators, target, key, desc);
    switch (arguments.length) {
        case 2: return decorators.reduceRight(function(o, d) { return (d && d(o)) || o; }, target);
        case 3: return decorators.reduceRight(function(o, d) { return (d && d(target, key)), void 0; }, void 0);
        case 4: return decorators.reduceRight(function(o, d) { return (d && d(target, key, o)) || o; }, desc);
    }
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
var di_1 = require('angular2/di');
var async_1 = require('angular2/src/facade/async');
var lang_1 = require('angular2/src/facade/lang');
var dom_adapter_1 = require('angular2/src/dom/dom_adapter');
var api_1 = require('../../api');
var compile_pipeline_1 = require('./compile_pipeline');
var view_loader_1 = require('angular2/src/render/dom/compiler/view_loader');
var compile_step_factory_1 = require('./compile_step_factory');
var change_detection_1 = require('angular2/change_detection');
var shadow_dom_strategy_1 = require('../shadow_dom/shadow_dom_strategy');
var pvm = require('../view/proto_view_merger');
/**
 * The compiler loads and translates the html templates of components into
 * nested ProtoViews. To decompose its functionality it uses
 * the CompilePipeline and the CompileSteps.
 */
var DomCompiler = (function (_super) {
    __extends(DomCompiler, _super);
    function DomCompiler(_stepFactory, _viewLoader, _useNativeShadowDom) {
        _super.call(this);
        this._stepFactory = _stepFactory;
        this._viewLoader = _viewLoader;
        this._useNativeShadowDom = _useNativeShadowDom;
    }
    DomCompiler.prototype.compile = function (view) {
        var _this = this;
        var tplPromise = this._viewLoader.load(view);
        return async_1.PromiseWrapper.then(tplPromise, function (el) { return _this._compileTemplate(view, el, api_1.ViewType.COMPONENT); }, function (e) {
            throw new lang_1.BaseException("Failed to load the template for \"" + view.componentId + "\" : " + e);
            return null;
        });
    };
    DomCompiler.prototype.compileHost = function (directiveMetadata) {
        var hostViewDef = new api_1.ViewDefinition({
            componentId: directiveMetadata.id,
            templateAbsUrl: null, template: null,
            styles: null,
            styleAbsUrls: null,
            directives: [directiveMetadata]
        });
        var template = dom_adapter_1.DOM.createTemplate('');
        dom_adapter_1.DOM.appendChild(dom_adapter_1.DOM.content(template), dom_adapter_1.DOM.createElement(directiveMetadata.selector));
        return this._compileTemplate(hostViewDef, template, api_1.ViewType.HOST);
    };
    DomCompiler.prototype.mergeProtoViewsRecursively = function (protoViewRefs) {
        return async_1.PromiseWrapper.resolve(pvm.mergeProtoViewsRecursively(protoViewRefs));
    };
    DomCompiler.prototype._compileTemplate = function (viewDef, tplElement, protoViewType) {
        var pipeline = new compile_pipeline_1.CompilePipeline(this._stepFactory.createSteps(viewDef), this._useNativeShadowDom);
        var compileElements = pipeline.process(tplElement, protoViewType, viewDef.componentId);
        return async_1.PromiseWrapper.resolve(compileElements[0].inheritedProtoView.build());
    };
    return DomCompiler;
})(api_1.RenderCompiler);
exports.DomCompiler = DomCompiler;
var DefaultDomCompiler = (function (_super) {
    __extends(DefaultDomCompiler, _super);
    function DefaultDomCompiler(parser, shadowDomStrategy, viewLoader) {
        _super.call(this, new compile_step_factory_1.DefaultStepFactory(parser, shadowDomStrategy), viewLoader, shadowDomStrategy.hasNativeContentElement());
    }
    DefaultDomCompiler = __decorate([
        di_1.Injectable(), 
        __metadata('design:paramtypes', [change_detection_1.Parser, shadow_dom_strategy_1.ShadowDomStrategy, view_loader_1.ViewLoader])
    ], DefaultDomCompiler);
    return DefaultDomCompiler;
})(DomCompiler);
exports.DefaultDomCompiler = DefaultDomCompiler;
